### PR TITLE
Improved `consolidate()` function and updated package website

### LINKED ISSUE #191: Improve `consolidate()` function
Currently consolidate has many more methods implemented, but this has slowed down the function and has made most the variables returned character class...

- To improve the speed and make the function more concise, we should reduce the number of for loops...
- We can resolve variables differently if they are not in character or messydt classes, keeping their original class...
- We were also getting some warnings related to NAs (specifically related to the NA remove argument for resolving dates with messydates), for now we suppress the warnings and sub NAs introduced in the resolve step by the first non NA value (coalesce). There were not many NAs returned when resolving to begin with but double check if this is the best way to proceed.

### LINKED ISSUE #194: Changes references from qID and qID_ref to treaty_ID and many_ID

### LINKED ISSUE #196: Update pkgdown sites to pkgdown 2.0.0
`{pkgdown}` 2.0.0. was [just released](https://pkgdown.r-lib.org/news/index.html). Some elements of the site configuration might need to be updated in consequence.

### LINKED ISSUE #199: Add "in_favour" method for resolving conflicts in `consolidate()` function in favor of certain datasets

### LINKED ISSUE #201: Find better ways to deal with issues related to ambiguous variable names in `consolidate()`
The function does not run properly when variable prefixes for different variables are the same (it assumes these are variables to be "consolidated"). Now we can deal with the issue several different ways, some (not mutually exclusive) options are below:

- Add an item on the PR template for many packages about making sure developers check if their databases work with consolidate before a version release
- Add more documentation/instructions for developers about variable naming practices expected for many packages
- Make the warnings in consolidate() more informative
- Make consolidate interactive (ask users) for when variable prefixes match but affixes are not the same (or the expected ones)

### LINKED ISSUE #202: Allow for key vector in consolidate() instead of single key variable
`dplyr::inner_join()` and `dplyr::full_join()` allow for a named vector to be passed to the `key` argument. We could implement this feature as well in our `consolidate()` function to allow for an easier joining of panel data for instance where individual observations are identified by a "country-year" pair. Here is an example:

``` r
consolidate(emperors, "any", "every", resolve = "min", key = c("ID" = "ID", "Year" = "Year"))
```

### PR DESCRIPTION
**Pull request by henriquesposito**
# Description

These patch changes improve `consolidate()` function by making it more concise and update the package website to work with the new `{pkgdown}` version release.

## Package

* Closed #194 by updating all remaining references from "qID" to "manyID"
* Updated package website
  * Closed #196 by updating elements that configure website to work properly
  * Updated '_pkgdown.yml' file to use bootstrap 5 template to build website

## Connection

* Updated `consolidate()` function
  * Closed #191 by making `consolidate()` function more concise and faster by removing redundant code lines
  * Fixed dates-related warnings by changing how `{messydates}` package is used to resolve dates
  * Updated how `consolidate()` substitutes missing observations with first non-missing observation from other datasets
  * Closed #201 by fixing how `consolidate()` detects variables to be resolved to avoid ambiguous variable matching
  *  Closed #202 by allowing for multiple key vectors to be declared as arguments for `consolidate()`
* Closed #199 by adding `favour()` (also `favor()`) function that re-orders datasets within a database

# Checklist:

- [x] PR form
  - [x] I have given this pull request an informative title
  - [x] Description above itemizes changes under subtitles, e.g. "## Collection""
  - [x] Any closed, fixed, or related issues are referenced and explained in the description above, e.g. "Fixed #0 by adding A"
  - [x] Package builds on my OS without issues
- [ ] PR checks all pass for latest commit
  - [x] Package builds on Mac
  - [x] Package builds on Windows
  - [x] Package builds on Linux
  - [ ] CodeCov check: Package improves or maintains good test coverage
  - [x] CodeFactor check: Package improves or maintains good style
- [x] Documentation
  - [x] Any new or modified functions or data have roxygen style documentation in their .R scripts
  - [x] Any longer functions are commented inline so that it is easier to debug in the future
  - [x] PR description above and the NEWS.md file are aligned
  - [x] DESCRIPTION file version is bumped by the appropriate increment (major, minor, patch)

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,20 @@
+# manydata 0.6.1
```
Comment (jhollway) [inline NEWS.md:1]: If this is the CRAN release, better start with a new (minor) version

File: DESCRIPTION [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Package: manydata
 Title: An R Portal for Ensembled Global Governance Data
-Version: 0.6.0
+Version: 0.6.1
 Date: 2021-12-03
```
Comment (jhollway) [inline DESCRIPTION:4]: Bump #minor. Also, Date needs updating...

File: DESCRIPTION [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Package: manydata
 Title: An R Portal for Ensembled Global Governance Data
```
Comment (jhollway) [inline DESCRIPTION:2]: Maybe doesn't need 'R' in the package title? Can't remember the CRAN regulations on this...

File: NAMESPACE [hunk 4]
```diff
@@ -7,6 +7,7 @@ export(data_contrast)
 export(data_source)
 export(extract_bilaterals)
 export(extract_multilaterals)
+export(favour)
```
Comment (jhollway) [inline NAMESPACE:10]: Does this need to be also re-exported as `favor` for our American cousins?

File: NEWS.md [hunk 5]
```diff
@@ -1,3 +1,20 @@
...
+* Updated package website
+  * Closed #196 by updating elements that configure website to work properly
+  * Updated `{pkgdown}` to use bootstrap 5 template
+* Closed #194 by updating all remaining references from qID to manyID
+
+## Connection
+
+* Updated `consolidate()` function
+  * Closed #191 by making `consolidate()` function more concise
+  * Fixed warnings by changing how ´{messydates}` package resolves dates
```
Comment (jhollway) [inline NEWS.md:14]: There's a front tick instead of a backtick on this line, and it confuses the syntax highlighter.

File: man/favour.Rd [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+% Please edit documentation in R/consolidate.R
+\name{favour}
+\alias{favour}
+\title{Favour datasets}
+\usage{
+favour(database, x)
+}
+\arguments{
+\item{database}{A many database}
+
```
Comment (jhollway) [inline man/favour.Rd:11]: This comment is for the logo -- it seems that it has shrunk considerably... Does this have any effect on how it shows on the website?

File: man/coalesce_compatible.Rd [hunk 7]
```diff
@@ -18,7 +18,7 @@ allowing for equality where one row has an element present
 and the other is missing the observation.
 }
 \examples{
```
Comment (jhollway) [inline man/coalesce_compatible.Rd:20]: This comment is for the logo -- it seems to have shrunk? Could this impact how it displays on the website/README?

**Review by jhollway — CHANGES_REQUESTED** [review]
Thanks @henriquesposito , I'm not finished going through it all, but here's a few comments to start with.

File: man/favour.Rd (hunk 6 above)
Comment (henriquesposito) [reply man/favour.Rd]: I see what you mean... I have changed the logo size specifications on the README file, but it seems now that the image quality become worse. @BBieri could you please take a look at the logo size and quality in the package README and website for manydata if you have a few minutes? Thank you.

File: DESCRIPTION [hunk 8]
```diff
@@ -31,11 +31,12 @@ License: CC BY 4.0
 URL: https://github.com/globalgov/manydata
 BugReports: https://github.com/globalgov/manydata/issues
 Depends:
-    R (>= 3.5.0),
-    tibble
+    R (>= 3.5.0)
```
Comment (jhollway) [inline DESCRIPTION:34]: Note that if we can avoid pointblank, we can go back to 3.4.0 -- if pointblank is only really used for manypkgs, maybe we can drop it?

File: DESCRIPTION [hunk 9]
```diff
@@ -1,7 +1,7 @@
 Package: manydata
-Title: An R Portal for Ensembled Global Governance Data
-Version: 0.6.0
-Date: 2021-12-03
+Title: An Portal for Ensembled Global Governance Data
```
Comment (jhollway) [inline DESCRIPTION:2]: A Portal

File: pkgdown/_pkgdown.yml [hunk 10]
```diff
@@ -51,22 +52,29 @@ navbar:
         - text: manytrade
           href: https://globalgov.github.io/manytrade/
 reference:
-  - title: "Obtaining and inspecting data in many packages"
-    desc: "These functions are to assist data analysts with the download of data on particular parts of global governance"
+  - title: "Downloading many packages and inspecting data"
```
Comment (jhollway) [inline pkgdown/_pkgdown.yml:55]: Phrasing could be confusing? Maybe 'many* packages' or 'manydata packages'?

File: pkgdown/_pkgdown.yml [hunk 11]
```diff
@@ -51,22 +52,29 @@ navbar:
...
+    desc: "These functions are to assist data analysts with the download of data on particular domains of global governance"
```
Comment (jhollway) [inline pkgdown/_pkgdown.yml:56]: Delete 'are to'

File: pkgdown/_pkgdown.yml [hunk 12]
```diff
@@ -51,22 +52,29 @@ navbar:
...
-    desc: "These functions assist data analysts with the analysis of data on particular parts of global governance"
+    desc: "These functions assist data analysts with the analysis of data on particular domains of global governance"
     contents:
       - consolidate
-      - coalesce_rows
       - coalesce_compatible
+      - favour
   - title: "Data wrangling with manydata"
-    desc: "These functions are to assist data analysts with the wrangling of data on particular parts of global governance"
+    desc: "These functions are to assist data analysts with the wrangling of data on particular domains of global governance"
```
Comment (jhollway) [inline pkgdown/_pkgdown.yml:67]: Delete 'are to'

Comment (jhollway) [reply pkgdown/_pkgdown.yml:67]: Maybe add a little more explanation? Something like: "Ensembled data adds special challenges to data wrangling. These functions assist analysts with wrangling such data"

File: pkgdown/_pkgdown.yml [hunk 13]
```diff
@@ -51,22 +52,29 @@ navbar:
...
-  - title: "Obtaining and inspecting data in many packages"
-    desc: "These functions are to assist data analysts with the download of data on particular parts of global governance"
+  - title: "Downloading many packages and inspecting data"
+    desc: "These functions are to assist data analysts with the download of data on particular domains of global governance"
     contents:
       - get_packages
       - starts_with("data_")
   - title: "Working with many databases"
-    desc: "These functions assist data analysts with the analysis of data on particular parts of global governance"
+    desc: "These functions assist data analysts with the analysis of data on particular domains of global governance"
```
Comment (jhollway) [inline pkgdown/_pkgdown.yml:61]: What's the difference between 'working with' and 'data wrangling'? Is there a more useful or informative way of categorising these functions?

File: pkgdown/_pkgdown.yml [hunk 14]
```diff
@@ -51,22 +52,29 @@ navbar:
...
       - coalesce_rows
       - recollect
       - reunite
       - repaint
+footer:
+  structure:
+    left: devs
+    right: panarchic
+  components:
+    devs: Developed by [James Hollway](https://jameshollway.com/).
```
Comment (jhollway) [inline pkgdown/_pkgdown.yml:79]: It's a team product. Just the panarchic link is fine here.

File: R/releases_plot.R [hunk 15]
```diff
@@ -8,103 +8,89 @@
 #' @importFrom tibble as_tibble
 #' @importFrom stats ave
 #' @importFrom stringr str_split str_remove
+#' @importFrom lubridate ymd ceiling_date floor_date
 #' @import ggplot2
-#' @import lubridate
```
Comment (jhollway) [inline R/releases_plot.R:12]: Is this the only function in the package with some of these dependencies, e.g. `ggplot2`? If so, do we really need the function here/on GitHub? If not, we can trim the package a little?

File: R/package_get.R [hunk 16]
```diff
@@ -130,20 +107,20 @@ get_packages <- function(pkg) {
...
     })
 
     repos <- tibble::as_tibble(dplyr::bind_rows(repos))
-    if(length(repos) < 2) {
+    if (length(repos) < 2) {
       stop("The download limit from GitHub has been reached.
-      To see all the packages in the many universe, please go to the following link:
+      To see all the packages in the many universe,
+      please go to the following link:
            https://github.com/globalgov")
```
Comment (jhollway) [inline R/package_get.R:118]: Does this stop message print all on one line or on separate lines? Maybe we want to print on separate lines?

File: R/extract.R [hunk 17]
```diff
@@ -0,0 +1,86 @@
+#' Extracting adjacency edgelist for international agreements
```
Comment (jhollway) [inline R/extract.R:1]: Does this only work on edgelists? Is 'edgelist' a known term for all those who might use the package? Could some users want to `extract_bilaterals()` just on the 'novelist' (as it were)?

File: R/consolidate.R [hunk 18]
```diff
@@ -211,6 +206,40 @@ compatible_rows <- function(x) {
...
+#' Favour datasets
+#' 
+#' @name favour
+#' @param database A many database
+#' @param x Would you like a dataset to be favoured over others?
+#' In this case the dataset selected becomes the reference for
+#' the first non NA value.
+#' If more than one dataset is declared,
+#' please add datasets increasing order of importance
+#' (.i.e. last dataset should be favoured over previous).
```
Comment (jhollway) [inline R/consolidate.R:218]: This is not a clear parameter description. The question posed suggests that a logical (TRUE/FALSE) answer is expected, but the example suggests that it's the name of one or more datasets?

File: R/consolidate.R [hunk 19]
```diff
@@ -239,24 +268,25 @@ r_coalesce <- function(other_variables, out, key) {
 #' @param other_variables A list of variables to be resolved
 #' @param out A dataframe
 #' @param key The ID column to collapse by. By default "many_ID"
+#' @import messydates
 #' @return The resolved dataframed or variable
 r_min <- function(other_variables, out, key) {
```
Comment (jhollway) [inline R/consolidate.R:273]: I find these functions really useful. Though I like the use of a prefix to avoid conflicts etc, I wonder whether in the greater `R` ecosystem `r_*` could be misunderstood? Maybe `res_*`, `rsv_*`, or `resolve_*` could work?

**Review by jhollway — CHANGES_REQUESTED** [review]
A few more ideas, thanks @henriquesposito

File: R/releases_plot.R (hunk 15 above)
Comment (henriquesposito) [reply R/releases_plot.R:12]: Yes, this is the case only for the `ggplot2` dependency ... Though I like having this function to complement the `get_packages()`function and help manydata to be the portal, I understand if you want to move it elsewhere. Just let me know.

File: man/favour.Rd (hunk 6 above)
Comment (BBieri) [reply man/favour.Rd]: I made this change following the update to a Bootstrap 5 based package website to be consistent with the `{tidyverse}` package logo sizes (see [the `{ggplot}` repository on GitHub](https://github.com/tidyverse/ggplot2/)). Let me know if you would like me to revert and keep the big logo @jhollway @henriquesposito.

File: R/package_get.R (hunk 16 above)
Comment (henriquesposito) [reply R/package_get.R:118]: It prints in separate lines

File: R/consolidate.R [hunk 20]
```diff
@@ -154,6 +148,16 @@ consolidate <- function(database,
   out
 }
 
+#' Selects a single dataset from a database
```
Comment (jhollway) [inline R/consolidate.R:151]: Can we add a bit more explanation here, for example where it comes from?

**Review by jhollway — CHANGES_REQUESTED** [review]

File: R/consolidate.R (hunk 20 above)
Comment (henriquesposito) [reply R/consolidate.R:151]: I added a bit more explanation in a details section. We cannot add a new description for functions that are reexported, we get a warning if we do so. Thank you.

File: R/releases_plot.R (hunk 15 above)
Comment (jhollway) [reply R/releases_plot.R:12]: Let's keep it for now, but it seems a candidate for being dropped at some stage.

File: R/extract.R [hunk 21]
```diff
@@ -0,0 +1,85 @@
+#' Extracting international agreements
+#' 
+#' @param membs A memberships dataset from one of the many packages
+#' @description This set of functions help extracting bilateral agreements or
+#' multilateral agreements from a membership database
+#' in the many packages universe.
+#' @name extraction
+NULL
+
+#' @name extraction
```
Comment (jhollway) [inline R/extract.R:10]: Isn't it `@rdname`?

File: R/extract.R [hunk 22]
```diff
@@ -0,0 +1,85 @@
+#' Extracting international agreements
+#' 
+#' @param membs A memberships dataset from one of the many packages
+#' @description This set of functions help extracting bilateral agreements or
+#' multilateral agreements from a membership database
+#' in the many packages universe.
```
Comment (jhollway) [inline R/extract.R:6]: A little more explanation of use cases? For example, should it be used before or after consolidating?

File: R/extract.R [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+  bilats <- dplyr::full_join(bilats2, bilats1,
+                             by = c("manyID", "Title", "Beg", "End")) %>%
+    dplyr::select(CountryID1, CountryID2, Title, Beg, End)
+  bilats
+}
+
+#' @name extraction
+#' @details The function helps extracting multilateral agreements
+#' from a membership database in the many packages universe.
+#' @importFrom dplyr filter
```
Comment (jhollway) [inline R/extract.R:57]: If you have `@import dplyr` above, you don't need `@importFrom dplyr ...` anywhere

File: R/data_report.R [hunk 24]
```diff
@@ -230,17 +235,13 @@ data_contrast <- function(pkg, database = NULL, dataset = NULL) {
 #' @name report
 #' @details `open_codebook()` opens the original codebook of the specified
 #' dataset to allow users to look up the original coding rules.
-#' Note that there is no original codebook for the genevar database
+#' Note that there is no original codebook for the GNEVAR database
```
Comment (jhollway) [inline R/data_report.R:238]: Do we explain what GNEVAR is anywhere? Do we really want to advertise that our contribution doesn't have a codebook?

File: R/consolidate.R [hunk 25]
```diff
@@ -396,3 +281,104 @@ r_random <- function(other_variables, out, key) {
...
+    fav <- database[dataset]
+    database[dataset] <- NULL
+    database <- append(fav, database)
+  }
+  database
+}
+
+#' @rdname favour
+#' @export
+favor <- favour
```
Comment (jhollway) [inline R/consolidate.R:384]: Is this the recommended way of managing function aliases?

**Review by jhollway — CHANGES_REQUESTED** [review]

File: DESCRIPTION [hunk 26]
```diff
@@ -1,7 +1,7 @@
 Package: manydata
-Title: An R Portal for Ensembled Global Governance Data
-Version: 0.6.0
-Date: 2021-12-03
+Title: A Portal for Ensembled Global Governance Data
+Version: 0.7.0
+Date: 2022-03-08
```
Comment (jaeltan) [inline DESCRIPTION:4]: Update date?

File: R/consolidate.R [hunk 27]
```diff
@@ -75,57 +70,55 @@ consolidate <- function(database,
   } else if (rows == "every") {
     out <- purrr::reduce(database, dplyr::inner_join, by = key)
   }
-
   # Step 2: Drop any unwanted variables
   cols <- match.arg(cols)
   all_variables <- unname(unlist(purrr::map(database, names)))
   if (cols == "every") {
     all_variables <- names(table(all_variables)[table(all_variables) == length(database)])
     out <- out %>% dplyr::select(all_of(key), starts_with(all_variables))
```
Comment (jaeltan) [inline R/consolidate.R:78]: do we need to add the 'all_of', 'starts_with' functions to importFrom list above?

File: R/consolidate.R [hunk 28]
```diff
@@ -396,3 +281,104 @@ r_random <- function(other_variables, out, key) {
   }
   out
 }
+
+#' Selects a single dataset from a database
+#' 
+#' @importFrom purrr pluck
+#' @return The selected database
```
Comment (jaeltan) [inline R/consolidate.R:288]: does it return a dataset or a database?

File: R/consolidate.R [hunk 29]
```diff
@@ -396,3 +281,104 @@ r_random <- function(other_variables, out, key) {
...
+    pairs[apply(t(compatico), 1, function(y) all(y)), ]
+  }
+}
+
+#' Favour datasets in a database
+#' 
+#' @name favour
+#' @param database A many database
+#' @param dataset The name of one, or more, 
+#' atasets within the database to be favoured over others.
```
Comment (jaeltan) [inline R/consolidate.R:356]: missing 'd' in 'datasets within the database'

File: R/wrangling_data.R [hunk 30]
```diff
@@ -11,23 +11,23 @@
 #' @import dplyr
 #' @source https://stackoverflow.com/questions/51428156/dplyr-mutate-transmute-drop-only-the-columns-used-in-the-formula
 #' @examples
-#' transmutate(mtcars, X = ifelse( vs, drat, wt ), Y = mpg*cyl)
+#' transmutate(emperors$wikipedia, Beggining = Beg)
```
Comment (jaeltan) [inline R/wrangling_data.R:14]: Beginning

File: vignettes/user.Rmd [hunk 31]
```diff
@@ -9,31 +9,22 @@ vignette: >
...
-(the main though not only R package repository), 
+Since some of these data packages are too big for CRAN, 
 we expect that their developers will instead choose
-to make their packages available on Github. 
-Github is huge though, either way it can be difficult to identify salient data. 
-We have tried to make this easy by allowing the reporting of all packages in the
-many packages universe, as well as the databases and datasets included in each
-of the 'many packages' - that we know of that currently pass our tests.
+to make their packages available on GitHub.
+To make it easier to identify of all packages in the
```
Comment (jaeltan) [inline vignettes/user.Rmd:23]: remove 'of'

File: vignettes/user.Rmd [hunk 32]
```diff
@@ -153,29 +145,29 @@ when emperors began their reign (removing any NAs).
...
-drawn using different datasets, let us examine how to select and collapse databases.
-`{manydata}` contains the `consolidate()` function which facilitates collapsing multiple
-datasets into a dataframe for analysis.
+Now that we have compared the data and looked at some of the different inferences drawn,
+let us examine how to select and consolidate databases.
 
-`consolidate()` collapses a set of datasets, or a database, from a 'many' package
-into a single dataset with some combination of the rows and columns.
-The function includes separate arguments for the rows and columns,
+The `consolidate()` function which facilitates consolidating a set of datasets, or a database,
```
Comment (jaeltan) [inline vignettes/user.Rmd:162]: remove 'which'

**Review by jaeltan — COMMENTED** [review]
Hi Henrique, I didn't have much to add except some minor spelling/wording issues. Thanks and good job!

File: R/consolidate.R (hunk 25 above)
Comment (henriquesposito) [reply R/consolidate.R:384]: If we want users to use the function and access the documentation, I think exporting is perhaps a more explicit way to do so. I previously tried to simply document as an alias and/or use the <- call (i.e. favour <- favor <- function ...), but had issues both approaches when trying to use and/access documentation for the function after loading the package locally. However, if you have any other ideas in how we should manage aliases please let me know and I will gladly change it. Thank you.

**Review by BBieri — APPROVED** [review]

**Comment by jhollway** [comment]
@jaeltan, can you please go through once more and explicitly approve this PR then?
@BBieri, would you like to add your point about vector keys for `consolidate()` to this PR?

File: NEWS.md [hunk 33]
```diff
@@ -1,3 +1,21 @@
+# manydata 0.7.0
+
+## Package
+
+* Updated package website
+  * Closed #196 by updating elements that configure website to work properly
+  * Updated `{pkgdown}` to use bootstrap 5 template
```
Comment (jhollway) [inline NEWS.md:7]: The package is not updated, it is the `_pkgdown.yaml` file that has been updated. Also, I'm still unclear what bootstrap 5 adds for us?

File: NEWS.md [hunk 34]
```diff
@@ -1,3 +1,21 @@
...
+* Closed #194 by updating all remaining references from qID to manyID
+
+## Connection
+
+* Updated `consolidate()` function
+  * Closed #191 by making `consolidate()` function more concise
+  * Fixed warnings by changing how `{messydates}` package resolves dates
+  * Updated how `consolidate()` substitutes NAs observations left with first non-missing observation
+  * Closed #201 by updating how `consolidate()` detects variables to be resolved
+* Closed #199 by adding `favour()` function that re-orders datasets within a database
```
Comment (jhollway) [inline NEWS.md:17]: Can we reword this to be a little more assertive/advertising? After all, this is the release that finally makes it on to CRAN... ;)

File: R/wrangling_data.R [hunk 35]
```diff
@@ -11,23 +11,23 @@
 #' @import dplyr
 #' @source https://stackoverflow.com/questions/51428156/dplyr-mutate-transmute-drop-only-the-columns-used-in-the-formula
 #' @examples
-#' transmutate(mtcars, X = ifelse( vs, drat, wt ), Y = mpg*cyl)
+#' transmutate(emperors$wikipedia, Beginning = Beg)
```
Comment (jhollway) [inline R/wrangling_data.R:14]: Can we add a line to show a before and after? e.g.

```
emperors$wikipedia
transmutate(emperors$wikipedia, Beginning = Beg)
```

File: R/extract.R [hunk 36]
```diff
@@ -0,0 +1,92 @@
+#' Extracting international agreements
```
Comment (jhollway) [inline R/extract.R:1]: I'm wondering whether "get" might be a more useful verb here than "extract", e.g. `get_bilaterals()`?

File: R/consolidate.R [hunk 37]
```diff
@@ -40,33 +39,29 @@
...
+#' consolidate(favour(emperors, "UNRV"), "every", "every",
+#' resolve = "coalesce", key = "ID")
+#' consolidate(emperors, "any", "every", resolve = "min", key = "ID")
+#' consolidate(emperors, "every", "any", resolve = "max", key = "ID")
+#' consolidate(emperors, "every", "every", resolve = "median", key = "ID")
 #' consolidate(emperors, "every", "every", resolve = "mean", key = "ID")
 #' consolidate(emperors, "every", "every", resolve = "random", key = "ID")
-#' consolidate(emperors, "every", "every", resolve = c(Beg = "min", End = "max"), key = "ID")
+#' consolidate(emperors, "every", "every",
+#' resolve = c(Beg = "min", End = "max"), key = "ID")
```
Comment (jhollway) [inline R/consolidate.R:57]: Maybe for these examples it would be useful to add the argument names. It's more verbose, but perhaps users would appreciate the examples being explicit?

**Review by jhollway — CHANGES_REQUESTED** [review]
A few more ideas here and there...

File: NEWS.md (hunk 33 above)
Comment (henriquesposito) [reply NEWS.md:7]: Yes, that is correct, thank you. I have changed the NEWS description to reflect such. As I understand, for the bootstrap 5 template upgrade, it reduces the chances of the website braking down and makes the CSS for the website a bit more programmable/flexible. However, @BBieri might be better equipped to tell you more about this upgrade.

File: R/extract.R (hunk 36 above)
Comment (henriquesposito) [reply R/extract.R:1]: I agree, this is also a bit more consistent with some of the other functions in the package as get_packages().

File: R/consolidate.R [hunk 38]
```diff
@@ -75,57 +76,55 @@ consolidate <- function(database,
   } else if (rows == "every") {
     out <- purrr::reduce(database, dplyr::inner_join, by = key)
   }
-
   # Step 2: Drop any unwanted variables
   cols <- match.arg(cols)
   all_variables <- unname(unlist(purrr::map(database, names)))
   if (cols == "every") {
     all_variables <- names(table(all_variables)[table(all_variables) == length(database)])
     out <- out %>% dplyr::select(all_of(key), starts_with(all_variables))
```
Comment (jhollway) [inline R/consolidate.R:84]: Need to prefix `all_of` and `starts_with` with package names

File: R/consolidate.R [hunk 39]
```diff
@@ -75,57 +76,55 @@ consolidate <- function(database,
   } else if (rows == "every") {
     out <- purrr::reduce(database, dplyr::inner_join, by = key)
   }
-
   # Step 2: Drop any unwanted variables
   cols <- match.arg(cols)
   all_variables <- unname(unlist(purrr::map(database, names)))
   if (cols == "every") {
     all_variables <- names(table(all_variables)[table(all_variables) == length(database)])
```
Comment (jhollway) [inline R/consolidate.R:83]: Long line?

File: R/consolidate.R [hunk 40]
```diff
@@ -40,33 +39,35 @@
...
+#' consolidate(database = emperors, rows = "every", cols = "every",
+#' resolve = "random", key = "ID")
+#' consolidate(database = emperors, rows = "every", cols = "every",
+#' resolve = c(Beg = "min", End = "max"), key = "ID")
 #' }
 #' @export
 consolidate <- function(database,
                         rows = c("any", "every"),
                         cols = c("any", "every"),
                         resolve = c("coalesce", "min", "max", "median", "mean", "random"),
```
Comment (jhollway) [inline R/consolidate.R:69]: Long line?

File: R/consolidate.R [hunk 41]
```diff
@@ -40,33 +39,35 @@
...
+#' resolve = "median", key = "ID")
+#' consolidate(database = emperors, rows = "every", cols = "every",
+#' resolve = "mean", key = "ID")
+#' consolidate(database = emperors, rows = "every", cols = "every",
+#' resolve = "random", key = "ID")
+#' consolidate(database = emperors, rows = "every", cols = "every",
+#' resolve = c(Beg = "min", End = "max"), key = "ID")
 #' }
 #' @export
 consolidate <- function(database,
```
Comment (jhollway) [inline R/consolidate.R:66]: Apparently this function has become overly complex (many if statements) -- can we reduce the number of switch points and/or unpack the function somehow?

File: R/consolidate.R [hunk 42]
```diff
@@ -154,245 +153,248 @@ consolidate <- function(database,
...
-#' @param key The ID column to collapse by. By default "many_ID"
-#' @return The resolved dataframed or variable
-r_coalesce <- function(other_variables, out, key) {
+resolve_coalesce <- function(other_variables, out, key) {
   for (var in other_variables) {
-    vars_to_combine <- startsWith(names(out), var)
+    vvars <- paste0("^", var, "$|^", var, "\\.")
+    vars_to_combine <- grepl(vvars, names(out))
     new_var <- dplyr::coalesce(!!!out[vars_to_combine])
     out <- out %>% dplyr::select(-dplyr::starts_with(var))
```
Comment (jhollway) [inline R/consolidate.R:161]: Are we re-exporting `%>%`? Otherwise we should try and avoid using this inside of functions

File: R/consolidate.R [hunk 43]
```diff
@@ -154,245 +153,248 @@ consolidate <- function(database,
...
     out[, var] <- new_var
   }
   if (length(other_variables) == 1) {
-    out <- dplyr::select(out, key, other_variables)
+    out <- dplyr::select(out, all_of(key), other_variables)
```
Comment (jhollway) [inline R/consolidate.R:165]: Need to prefix `all_of` with package name

File: R/consolidate.R [hunk 44]
```diff
@@ -154,245 +153,248 @@ consolidate <- function(database,
...
     a <- dplyr::coalesce(!!!out[vars_to_combine])
     new_var <- ifelse(is.na(new_var), a, new_var)
-    out <- out %>% dplyr::select(-dplyr::starts_with(var))
+    out <- dplyr::select(out, -dplyr::starts_with(var))
+    new_var
     out[, var] <- new_var
   }
   if (length(other_variables) == 1) {
-    out <- dplyr::select(out, key, other_variables)
+    out <- dplyr::select(out, all_of(key), other_variables)
```
Comment (jhollway) [inline R/consolidate.R:191]: Need to prefix `all_of` with package name

File: R/consolidate.R [hunk 45]
```diff
@@ -154,245 +153,248 @@ consolidate <- function(database,
...
```
Comment (jhollway) [inline R/consolidate.R:217]: Need to prefix `all_of` with package name

File: R/consolidate.R (hunk 41 above)
Comment (henriquesposito) [reply R/consolidate.R:66]: I reduced the complexity from 27 to 16 by adding helper functions to resolve variables for unique or multiple arguments. I can reduce this complexity further, but I have the impression that we would have too many layers of helper functions (i.e. helper functions that call other helper functions). Would you like me to reduce it more? Thank you.

File: R/consolidate.R [hunk 46]
```diff
@@ -1,19 +1,18 @@
 #' Consolidate database into a single dataset
 #'
-#' This function collapses a set or database of many datasets
-#' into a single dataset with some combination of the
-#' rows, columns, and observations of the parent datasets.
+#' This function consolidates a set of datasets in a 'many* package' database
```
Comment (jaeltan) [inline R/consolidate.R:3]: Should we add a note about the naming of variables? So that users don't run into the same problems that we did

File: R/consolidate.R [hunk 47]
```diff
@@ -26,373 +25,388 @@
...
+#' Favour datasets in a database
+#'
+#' @name favour
+#' @param database A many database
+#' @param dataset The name of one, or more, datasets within the database
+#' to be favoured over others.
+#' @details The dataset declared becomes the reference for
+#' the first non NA value.
+#' If more than one dataset is declared,
+#' please add datasets increasing order of importance
```
Comment (jaeltan) [inline R/consolidate.R:387]: minor rephrase: 'please list datasets in increasing order of importance'

**Review by jaeltan — COMMENTED** [review]
Thanks Henrique for all the hard work on manydata! I just made a few minor comments. Also I pushed the changes to the cheatsheet files to the /inst folder but the readme seems to use the one from the /man/figures folder, should I move the edited ones over to the /man/figures folder and remove the ones from /inst?

File: R/consolidate.R (hunk 46 above)
[the remark this answers is not in the document]
Comment (henriquesposito) [reply R/consolidate.R:3]: Thank you for the comment Jael, but I actually already fixed the issue with ambiguous variable names. I do think that maybe, in manypkgs, we could add a sentence on the preparation script about how having one letter variable names is not recommend. What do you think?

[the remark this answers is not in the document]
Comment (jaeltan) [reply R/consolidate.R:3]: I think that's a good idea, thank you!

File: R/treaties_get.R [hunk 48]
```diff
@@ -0,0 +1,93 @@
+#' Get international treaties
+#'
+#' @description Some datasets in the membership databases across the
+#' 'many* packages' (e.g. manyenviron) contain a myriad of information on
+#' international treaties governing an international domain.
+#' The get treaty functions help researchers retrieve all bilateral
+#' agreements, or multilateral agreements, from these datasets.
+#' Researchers can, for example, use `get_bilaterals()` to retrieve
```
[the remark this answers is not in the document]
Comment (jaeltan) [reply R/treaties_get.R:8]: I think it's fine, I just wanted to clarify that `get_bilaterals()` returns the countries but `get_multilaterals()` doesn't

**Review by BBieri — APPROVED** [review]

File: R/consolidate.R (hunk 41 above)
Comment (henriquesposito) [reply R/consolidate.R]: I reduced the complexity below 15 just in case. Thank you.

**Review by jhollway — APPROVED** [review]
